### PR TITLE
policy: Add selector cache

### PR DESCRIPTION
**Pull request by jrajahalme**
Add an identity selector cache that does all the selector matching
necessary and caches the results.

This cache also gets updates from the kvstore on added and deleted
identities, and updates the selections accordingly.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7539)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/identity/cache/cache.go [hunk 1]
```diff
@@ -103,26 +102,64 @@ type identityWatcher struct {
...
+	if gi, ok := event.Key.(globalIdentity); ok {
+		id := identity.NumericIdentity(event.ID)
+		lbls := gi.LabelArray()
+		if event.Typ == kvstore.EventTypeCreate {
+			added[id] = lbls
+		} else {
+			deleted[id] = lbls
+		}
+		return true
+	} else {
```
[the remark this answers is not in the document]
Comment (jrajahalme) [reply pkg/identity/cache/cache.go]: All of the above fixed for the next push.

**Comment by jrajahalme** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

File: pkg/identity/cache/cache.go [hunk 2]
```diff
@@ -103,26 +102,63 @@ type identityWatcher struct {
...
 			select {
 			case event := <-events:
-
+				// Collect first added and deleted labels
 				switch event.Typ {
 				case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
-					owner.TriggerPolicyUpdates(true, "one or more identities created or deleted")
-
+					if collectEvent(event, added, deleted) {
+						break
```
Comment (joestringer) [inline pkg/identity/cache/cache.go]: What does this `break` break?

File: pkg/identity/cache/cache.go [hunk 3]
```diff
@@ -103,26 +102,63 @@ type identityWatcher struct {
...
 				case kvstore.EventTypeModify:
 					// Ignore modify events
 				}
 
 			case <-w.stopChan:
 				return
 			}
+
+			// see if there is more, but do not wait nor stop
+			select {
```
Comment (joestringer) [inline pkg/identity/cache/cache.go]: This only collects one more event. I suspect that you want something similar to the following incremental patch?

```
@@ -145,20 +145,23 @@ func (w *identityWatcher) watch(owner IdentityAllocatorOwner, events allocator.A
                        }
 
                        // see if there is more, but do not wait nor stop
-                       select {
-                       case event := <-events:
-                               // Collect more added and deleted labels
-                               switch event.Typ {
-                               case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
-                                       collectEvent(event, added, deleted)
-                               case kvstore.EventTypeModify:
-                                       // Ignore modify events
+                       for {
+                               select {
+                               case event := <-events:
+                                       // Collect more added and deleted labels
+                                       switch event.Typ {
+                                       case kvstore.EventTypeCreate, kvstore.EventTypeDelete:
+                                               collectEvent(event, added, deleted)
+                                       case kvstore.EventTypeModify:
+                                               // Ignore modify events
+                                       }
+                               default:
                                }
-
-                       default:
-                               // Issue collected updates
-                               owner.UpdateIdentities(added, deleted)
                        }
+
+                       // Issue collected updates
+                       owner.UpdateIdentities(added, deleted)
+
                }
        }()
 }
```

(Seems like potentially a good target for a unit test to check that batching works correctly..)

File: pkg/policy/selectorcache.go [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+type labelIdentitySelector struct {
+	selector         api.EndpointSelector
+	key              string
+	users            map[CachedSelectionUser]struct{}
+	selections       unsafe.Pointer
+	cachedSelections map[identity.NumericIdentity]struct{}
+}
+
+//
+// CachcedSelector implemenentation (== Public API)
```
Comment (joestringer) [inline pkg/policy/selectorcache.go]:
```suggestion
// CachedSelector implementation (== Public API)
```

File: pkg/policy/selectorcache.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+	return *(*[]identity.NumericIdentity)(atomic.LoadPointer(&l.selections))
+}
+
+// String returns the map key for this selector
+func (l *labelIdentitySelector) String() string {
+	return l.key
+}
+
+//
+// IdentitySelector implemenentation (== internal API)
```
Comment (joestringer) [inline pkg/policy/selectorcache.go]:
```suggestion
// IdentitySelector implementation (== internal API)
```

File: pkg/identity/cache/cache.go [hunk 6]
```diff
@@ -103,26 +102,63 @@ type identityWatcher struct {
...
+		id := identity.NumericIdentity(event.ID)
+		lbls := gi.LabelArray()
+		if event.Typ == kvstore.EventTypeCreate {
+			added[id] = lbls
+		} else {
+			deleted[id] = lbls
+		}
+		return true
+	}
+	log.Warningf("Ignoring unknown identity type '%s': %+v",
```
Comment (joestringer) [inline pkg/identity/cache/cache.go]: Please use the `log.WithFields(...).Warning(...)` style across this PR. For log management systems and even our CI, this makes errors easier to track. We now do things in the CI like counting the number of warnings with a unique message to figure out the most popular problems, so if the content of the log message has variable contents it can mess with that type of operation over the logs.

**Review by joestringer — CHANGES_REQUESTED** [review]
This is really slick! Core logic looks all good, I have just one note on event batching below. You might consider also putting a `Trigger` there too, just to put a bound on how often we cause churn in the subscribers.

I realize that the SelectorCache is written in such a way that concurrent updates will not cause problems (a mutex protects write + atomic read pointer to immutable state), but it'd be nice to see unit tests with concurrent writes and concurrent reads + writes.

File: pkg/identity/cache/cache.go (hunk 2 above)
Comment (joestringer) [reply pkg/identity/cache/cache.go]: (to answer my question, it's effectively a no-op; it breaks the switch which will exit immediately following this `collectEvent()` call anyway.)

Comment (joestringer) [reply pkg/identity/cache/cache.go]: I think this is intended to have a label like `waitForEvent:` above the select, and to `break waitForEvent`

File: pkg/policy/selectorcache.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+// IdentitySelector. Users of the SelectorCache take care of creating
+// IdentitySelectors as needed by identity policies. The set of
+// IdentitySelectors is read locked during an IdentityPolicy update so
+// that the the policy is always updated using a coherent set of
+// cached selections.
+//
+// IdentitySelector is used as a map key, so it must not be implemented by a
+// map, slice, or a func, or a runtime panic will be triggered. In all
+// cases below IdentitySelector is being implemented by structs.
+type IdentitySelector interface {
```
Comment (joestringer) [inline pkg/policy/selectorcache.go]: By the way I think this can be hidden, `identitySelector` since the interface isn't used outside this package.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Some raciness in the augmented unit tests, adjusting and re-testing.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/identity/cache/cache.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/identity/cache/cache.go]: Thanks for spotting this! Fixed and added unit tests to make sure all of this is covered and works as intended :-)

File: pkg/identity/cache/cache.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/identity/cache/cache.go]: Fixed!

File: pkg/policy/selectorcache.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Fixed

File: pkg/policy/selectorcache.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Fixed

File: pkg/identity/cache/cache.go (hunk 6 above)
Comment (jrajahalme) [reply pkg/identity/cache/cache.go]: Done, mostly. For this one I did not, as this is something that should happen only if something is severely wrong with the source code, and could not readily figure out if there are suitable fields to use. Should I just invent new ones in this case?

File: pkg/policy/selectorcache.go (hunk 7 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Will do for the next round.

File: pkg/labels/labels.go [hunk 8]
```diff
@@ -458,11 +458,17 @@ func (l Labels) ToSlice() []Label {
 	return labels
 }
 
-// LabelArray returns the labels as label array
+// LabelArray returns the labels as label array, sorted by the key
 func (l Labels) LabelArray() LabelArray {
```
Comment (ianvernon) [inline pkg/labels/labels.go]: Add a unit test to ensure that the labels are correctly sorted.

File: pkg/labels/array.go [hunk 9]
```diff
@@ -38,10 +42,19 @@ func ParseSelectLabelArray(labels ...string) LabelArray {
...
+			array[i] = lbl
+			i++
+		}
 	}
-	return array
+	return array[:i]
+}
+
+// NewLabelArrayFromSortedList returns labels based on the output of SortedList()
+func NewLabelArrayFromSortedList(list string) LabelArray {
```
Comment (ianvernon) [inline pkg/labels/array.go]: Add a unit test here as well.

File: pkg/labels/array.go [hunk 10]
```diff
@@ -38,10 +42,19 @@ func ParseSelectLabelArray(labels ...string) LabelArray {
 // ParseLabelArrayFromArray converts an array of strings as labels and returns a LabelArray
 func ParseLabelArrayFromArray(base []string) LabelArray {
 	array := make([]Label, len(base))
-	for i := range base {
-		array[i] = ParseLabel(base[i])
+	i := 0
+	for _, v := range base {
+		if lbl := ParseLabel(v); lbl.Key != "" {
```
Comment (ianvernon) [inline pkg/labels/array.go]: Should we warn via a log message if the key is equal to an empty string? When would such a case happen?

File: pkg/identity/cache/allocator.go [hunk 11]
```diff
@@ -33,22 +33,35 @@ import (
 
 // globalIdentity is the structure used to store an identity in the kvstore
 type globalIdentity struct {
-	labels.Labels
+	labels.LabelArray
```
Comment (ianvernon) [inline pkg/identity/cache/allocator.go]: All of these commits around labels-related changes seem to be better for another PR outside of this selector cache implementation. Can you factor this commit and the following into a separate PR?

* 7f000d77731aeab68f556826517239635a1af0d9
* 0ce43c1f569b8007050c56eff8b67b2718d05a38
* 90f4bea8b7a998658bf0350762fc0df5d19e9e74

File: pkg/identity/cache/allocator.go [hunk 12]
```diff
@@ -33,22 +33,35 @@ import (
...
-	return kvstore.Encode(gi.SortedList())
+	str := ""
+	for _, l := range gi.LabelArray {
+		// We don't care if the values already have a '=' since this method is
+		// only used to calculate a SHA256Sum
+		//
+		// We absolutely care that the final character is a semi-colon.
+		// Identity allocation in the kvstore depends on this (see
+		// kvstore.prefixMatchesKey())
+		str += fmt.Sprintf(`%s:%s=%s;`, l.Source, l.Key, l.Value)
```
Comment (ianvernon) [inline pkg/identity/cache/allocator.go]: @tgraf PTAL here - this looks good to me, but want to make sure you get sign-off on this as this is critical functionality

File: pkg/identity/cache/allocation_test.go [hunk 13]
```diff
@@ -100,17 +104,120 @@ func (e *IdentityAllocatorConsulSuite) SetUpTest(c *C) {
...
+		case nid, ok := <-d.updated:
+			if (!ok) {
+				// updates channel closed
+				return 0
+			}
+			rounds++
+			if nid == target {
+				return rounds
+			}
+		case <-time.After(60 * time.Second):
```
Comment (ianvernon) [inline pkg/identity/cache/allocation_test.go:161]: This timeout would occur only for the given `select` statement (i.e., if nothing is sent to `d.updated`) vs. after an entire duration of 60 seconds for `target` to be reached for all iterations of receiving off of `d.updated` - is this intentional? Should the time be declared outside of the `for` loop, and then have `case <- (60 seconds since time outside of loop)`?

File: pkg/identity/cache/allocation_test.go [hunk 14]
```diff
@@ -212,6 +343,10 @@ func (ias *IdentityAllocatorSuite) TestLocalAllocationr(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(released, Equals, true)
 
+	// KV-store still holds on to the key, so it is not deleted yet via KV-store events
+	// This may be racy, as timing here depends on scheduling.
+	c.Assert(owner.GetIdentity(id.ID), checker.DeepEquals, lbls1.LabelArray())
```
Comment (ianvernon) [inline pkg/identity/cache/allocation_test.go:349]: you can use `testutils.WaitUntil` here for functionality that is ran in the background but should return after some amount of time.

File: pkg/identity/cache/cache.go [hunk 15]
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
...
+			// Un-delete the added ID if previously
+			// 'deleted' so that collected events can be
+			// processed in any order.
+			if _, exists := deleted[id]; exists {
+				delete(deleted, id)
+			}
+			added[id] = gi.LabelArray
+			return true
+		}
+		log.Warningf("collectEvent: Ignoring unknown identity type '%s': %+v",
```
Comment (ianvernon) [inline pkg/identity/cache/cache.go:120]: Use `log.WithFields` here and have fields for the type of the key / the key itself.

File: pkg/identity/cache/cache.go [hunk 16]
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
 	stopChan chan bool
 }
 
+func collectEvent(event allocator.AllocatorEvent, added, deleted IdentityCache) bool {
```
Comment (ianvernon) [inline pkg/identity/cache/cache.go:106]: Even though this function is not exported, I think this should be unit tested, as it is much easier to unit test than the  `watch` logic.

File: pkg/identity/cache/cache.go [hunk 17]
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
...
+				delete(deleted, id)
+			}
+			added[id] = gi.LabelArray
+			return true
+		}
+		log.Warningf("collectEvent: Ignoring unknown identity type '%s': %+v",
+			reflect.TypeOf(event.Key), event.Key)
+		return false
+	}
+	// Reverse an add when subsequently deleted
```
Comment (ianvernon) [inline pkg/identity/cache/cache.go:124]: This code confused me, as it is implicit that we only reach this event if it is a `EventTypeDelete`, as the code always returns a value if the if-statement on line 109 evaluates to true. Can you make this a switch statement based off of the type of `event.Typ`? It makes the code clearer.

File: pkg/identity/cache/cache.go (hunk 16 above)
Comment (ianvernon) [inline pkg/identity/cache/cache.go:106]: Add a comment indicating what the bool represents.

File: pkg/identity/cache/cache.go [hunk 18]
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
...
+						collectEvent(event, added, deleted)
+					default:
+						// Ignore modify events
+					}
+				default:
+					// No more events available without blocking
+					break More
+				}
 			}
+			// Issue collected updates
```
Comment (ianvernon) [inline pkg/identity/cache/cache.go:190]: Should there be a `select` here if `w.stopChan` is closed? That way we don't have to issue identity updates to the owner.

File: pkg/identity/cache/cache.go [hunk 19]
```diff
@@ -103,26 +103,92 @@ type identityWatcher struct {
...
+				case <-w.stopChan:
+					return
 				}
+			}
 
-			case <-w.stopChan:
-				return
+		More:
+			for {
+				// see if there is more, but do not wait nor stop
```
Comment (ianvernon) [inline pkg/identity/cache/cache.go:171]: Can you elaborate on the rationale between having separate loops for `First` and `More` and why we care if `w.stopChan` is closed in `First` vs. not in `More`? Shouldn't we always stop if stopped?

File: pkg/policy/selectorcache.go [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+
+	"github.com/sirupsen/logrus"
+)
+
+// CachedSelector represents an identity selector owned by the selector cache
+type CachedSelector interface {
+	// GetSelections returns the cached set of numeric identities
+	// selected by the CachedSelector.  The retuned slice must NOT
+	// be modified, as it is shared among multiple users.
+	GetSelections() []identity.NumericIdentity
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:36]: Maybe wrap `[]identity.NumericIdentity` around a type and make the slice private? Then we can have an iterator function for it which returns the identities to those who need to iterate over it, and we can ensure that nothing can modify it without having to specify it here. This makes the modification impossible from outside users.

File: pkg/policy/selectorcache.go [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+	// Used as a map key.
+	String() string
+}
+
+// CachedSelectionUser inserts selectors into the cache and gets update
+// callbacks whenever the set of selected numeric identities change for
+// the CachedSelectors pushed by it.
+type CachedSelectionUser interface {
+	// IdentitySelectionUpdated implementations MUST NOT call back
+	// to selector cache while executing this function!
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:48]: What is the reason for this requirement?

File: pkg/policy/selectorcache.go [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+	// kv-store and the local identity facility via our
+	// UpdateIdentities() function.
+	idCache cache.IdentityCache
+
+	// map key is the string representation of the selector being cached.
+	selectors map[string]IdentitySelector
+}
+
+// NewSelectorCache creates a new SelectorCache.
+func NewSelectorCache() *SelectorCache {
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go]: This could be made private since it's not used anywhere outside of this package.

File: pkg/policy/selectorcache.go [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+		idCache:   cache.GetIdentityCache(),
+		selectors: make(map[string]IdentitySelector),
+	}
+}
+
+var (
+	// selectorCache is the global selector cache. Additional ones
+	// are only created for testing.
+	selectorCache  = NewSelectorCache()
+	emptySelection = make([]identity.NumericIdentity, 0)
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go]: What does this global do? Please add a comment here.

File: pkg/policy/selectorcache.go [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+// CachedSelector implementation (== Public API)
+//
+// No locking needed.
+//
+
+// GetSelections returns the set of numeric identities currently
+// selected.  The cached selections can be concurrently updated. In
+// that case GetSelections() will return either the old or new version
+// of the selections. If the old version is returned, the user is
+// guaranteed to receive a notification including the update.
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:138]: including the update of the new version?

File: pkg/policy/selectorcache.go [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+	// are only created for testing.
+	selectorCache  = NewSelectorCache()
+	emptySelection = make([]identity.NumericIdentity, 0)
+)
+
+type labelIdentitySelector struct {
+	selector         api.EndpointSelector
+	key              string
+	users            map[CachedSelectionUser]struct{}
+	selections       unsafe.Pointer
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go]: add a comment indicating that this will always be a slice of `identity.NumericIdentity`

File: pkg/policy/selectorcache.go [hunk 26]
```diff
@@ -0,0 +1,336 @@
...
+//
+// lock must be held
+func (l *labelIdentitySelector) updateSelections() {
+	selections := make([]identity.NumericIdentity, len(l.cachedSelections))
+	i := 0
+	for nid := range l.cachedSelections {
+		selections[i] = nid
+		i++
+	}
+	// Sort the numeric identities so that the map iteration order
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:193]: I don't understand this comment - map iteration order is random, so why do we need to sort?

File: pkg/policy/selectorcache.go [hunk 27]
```diff
@@ -0,0 +1,336 @@
...
+	labelIdentitySelector
+	// fqdnSelections map[string]identity.NumericIdentity // identity.String(): "cidr:1.1.1.1" -> identity of 1.1.1.1
+}
+
+// AddIdentitySelector adds the given api.EndpointSelector in to the
+// selector cache. If an identical EndpointSelector has already been
+// cached, the corresponding CachedSelector is returned, otherwise one
+// is created and added to the cache.
+func (sc *SelectorCache) AddIdentitySelector(user CachedSelectionUser, selector api.EndpointSelector) (cachedSelector CachedSelector, added bool) {
+	key := selector.LabelSelector.String()
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:216]: Add a comment that that `String` may not always be identical for the case of requirements being in different order for `LabelSelector`.

File: pkg/policy/selectorcache.go [hunk 28]
```diff
@@ -0,0 +1,336 @@
...
+	}
+	// Create the immutable slice representation of the selected
+	// numeric identities
+	newIDSel.updateSelections()
+
+	// Note: No notifications are sent for the existing
+	// identities. Caller must use GetSelections() to get the
+	// current selections after adding a selector. This way the
+	// behavior is the same between the two cases here (selector
+	// is already cached, or is a new one).
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:253]: We will have to account for this behavior for incremental updates, though. We will have to figure out how to communicate this downstream to identity policies in an elegant manner.

File: pkg/policy/selectorcache.go [hunk 29]
```diff
@@ -0,0 +1,336 @@
...
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
+
+	// Update idCache so that newly added selectors get
+	// prepopulated with all matching numeric identities.
+	for numericID := range deleted {
+		if _, exists := sc.idCache[numericID]; exists {
+			delete(sc.idCache, numericID)
+		} else {
+			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warningf("UpdateIdentities: Deleting a non-existing identity")
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go]: Warningf --> Warning

File: pkg/policy/selectorcache.go [hunk 30]
```diff
@@ -0,0 +1,336 @@
...
+		}
+	}
+	for numericID, lbls := range added {
+		if _, exists := sc.idCache[numericID]; exists {
+			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warningf("UpdateIdentities: Adding an existing identity")
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go]: Warningf --> Warning

File: pkg/policy/selectorcache.go [hunk 31]
```diff
@@ -0,0 +1,336 @@
...
+					}
+				}
+			}
+			if len(dels)+len(adds) > 0 {
+				idSel.updateSelections()
+				idSel.notifyUsers(adds, dels)
+			}
+		case *fqdnSelector:
+			// TODO
+		}
```
Comment (ianvernon) [inline pkg/policy/selectorcache.go:319]: Should we have a default case with an error? Not sure if this would ever happen, but could be helpful if someone adds another selector type in the future.

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: pkg/identity/cache/cache.go (hunk 6 above)
Comment (joestringer) [reply pkg/identity/cache/cache.go]: (Sorry, missed this earlier) @jrajahalme if you think there's a case to reuse across multiple packages, be sure to add new fields to `pkg/logging/logfields/`, otherwise I'm fine with just defining the parameter titles using direct strings here.

Given the `%+v`, I think it's important to do so with this particular message.

File: pkg/identity/cache/allocator.go (hunk 11 above)
Comment (jrajahalme) [reply pkg/identity/cache/allocator.go]: @ianvernon Looking into this, but it seems that if we split this up at least one more of the remaining commits would need to be combined with these. This is due to the new function `identity.NewIdentityFromLabelArray()` introduced here only being used in the next commit. Will post a separate PR for the first 4 commits.

File: pkg/labels/labels.go (hunk 8 above)
Comment (jrajahalme) [reply pkg/labels/labels.go]: Added in #7815

File: pkg/labels/array.go (hunk 9 above)
Comment (jrajahalme) [reply pkg/labels/array.go]: Added in #7815

File: pkg/labels/array.go (hunk 10 above)
Comment (jrajahalme) [reply pkg/labels/array.go]: The SortedList format terminates in a ';', which will produce an empty string when split. That's the only case I know of.

File: pkg/identity/cache/allocator.go (hunk 11 above)
Comment (jrajahalme) [reply pkg/identity/cache/allocator.go]: Done (#7815)

File: pkg/identity/cache/allocator.go (hunk 12 above)
Comment (jrajahalme) [reply pkg/identity/cache/allocator.go]: I factored this piece out so that this is only in one place, should be more straightforward to review!

File: pkg/identity/cache/allocation_test.go (hunk 13 above)
Comment (jrajahalme) [reply pkg/identity/cache/allocation_test.go:161]: This is a testing utility, so the exact timeout semantics is not very significant. I kept the function as is and changed the function comment to end in:
```
// Returns 0 in case of 'd.updated' channel is closed or
// nothing is received from that channel in 60 seconds.
```

File: pkg/identity/cache/allocation_test.go (hunk 14 above)
Comment (jrajahalme) [reply pkg/identity/cache/allocation_test.go:349]: I don't know the exact timing, but the KV-store should not immediately release the key. This test is for testing that this is the case, not for the fact that the key is eventually deleted from the KV-store.

File: pkg/identity/cache/cache.go (hunk 19 above)
Comment (jrajahalme) [reply pkg/identity/cache/cache.go:171]: The intent is to block or stop before the first event is received, and then pool as many other events together with the first one to amortize processing cost. The second loop does not block, and if we got an event, it was received before the stop channel was closed, so it should be processed. The stop channel will be tested again as soon as the first received event (together with any other events) is processed.

We could add check for the closure of the stop channel to the second loop, but IMO we still should process all events received before the stop channel was closed.

File: pkg/policy/selectorcache.go (hunk 20 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go:36]: Let's see how we end up using this. Passing a slice onto which the numeric identities are appended by this function would be much simpler, if that satisfies the use case. But likely we'll be creating bpf map entries, one per numeric identity, and in that case iterator would be a good fit.

File: pkg/policy/selectorcache.go (hunk 21 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go:48]: To avoid recursive/double locking. This function (callback) is called with the cache lock held, and calling any selector cache "API" function will panic.

File: pkg/policy/selectorcache.go (hunk 22 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Right, will do.

File: pkg/policy/selectorcache.go (hunk 23 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Don't know how often this will be used, but is more cache efficient if we don't need multiple copies of empty slices (both slice header and the backing memory). We could maybe expand this to only keep one copy of any kind of a slice. We already keep them sorted, so it might be possible to build an index to them. I added this comment and changed the definition slightly as a nil slice should work as well:
```
	// Empty slice of numeric identities used for all selectors that select nothing
	emptySelection []identity.NumericIdentity
```

File: pkg/policy/selectorcache.go (hunk 24 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go:138]: Right, there will be a notification of for the change that caused the new version to be created, but as updates are coalesced by the next time the user calls `GetSelections()` it may already be updated with further changes.

File: pkg/policy/selectorcache.go (hunk 25 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: done!

File: pkg/policy/selectorcache.go (hunk 26 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go:193]: Meaning that the order of the numeric identities in the slice will be the same regardless the what iteration order of the map happens to be. Makes testing and debugging easier.

File: pkg/policy/selectorcache.go (hunk 27 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go:216]: Added this:
```
	// The key returned here may be different for equivalent
	// labelselectors, if the selector's requirements are stored
	// in different orders. When this happens we'll be tracking
	// essentially two copies of the same selector.
```

File: pkg/policy/selectorcache.go (hunk 28 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go:253]: We may need to add a revision number to both the `GetSelections()` and incremental updates to properly sequence them on the user side.

File: pkg/policy/selectorcache.go (hunk 29 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Thnx!

File: pkg/policy/selectorcache.go (hunk 30 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: x2

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/selectorcache.go (hunk 7 above)
Comment (jrajahalme) [reply pkg/policy/selectorcache.go]: Done!

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/identity/cache/cache.go (hunk 15 above)
Comment (joestringer) [reply pkg/identity/cache/cache.go:120]: This one still doesn't appear to be resolved ^

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/identity/cache/cache.go (hunk 16 above)
Comment (ianvernon) [reply pkg/identity/cache/cache.go:106]: @jrajahalme ^^

File: pkg/identity/cache/cache.go (hunk 15 above)
Comment (ianvernon) [reply pkg/identity/cache/cache.go:120]: @jrajahalme ^^

File: pkg/identity/cache/cache.go (hunk 17 above)
Comment (ianvernon) [reply pkg/identity/cache/cache.go:124]: @jrajahalme ^^

File: pkg/policy/selectorcache.go (hunk 20 above)
Comment (ianvernon) [reply pkg/policy/selectorcache.go:36]: 👍 , we can always add this later.

File: pkg/policy/selectorcache.go (hunk 31 above)
Comment (ianvernon) [reply pkg/policy/selectorcache.go:319]: @jrajahalme thoughts on this? ^ we may not even need this switch statement in the FQDN case since there is no logical way of corresponding a CIDR identity with a given FQDN by virtue of the fields of the identity itself at this current time - this relation has to come from the DNS proxy.

**Review by ianvernon — APPROVED** [review]
Nothing blocking - this is solid enough for now that we can merge it, especially since this is not being plumbed yet throughout policy calculation. We will tune it as needed to meet our requirements. Ship it!

**Comment by ianvernon** [comment]
@joestringer said we can merge this offline, so I am merging. :shipit:
